lib/apiservers/engine/backends/copy.go

```diff
@@ -0,0 +1,514 @@
+	readerMap := NewArchiveStreamReaderMap(mounts)
+
+	readers, err := readerMap.ReadersForSourcePath(c.containerProxy, vc.ContainerID, path)
+	if err != nil {
+		log.Errorf("Errors getting readers for export: %s", err.Error())
+		return nil, err
+	}
+
+	log.Infof("Got %d archive readers", len(readers))
+	finalTarReader := io.MultiReader(readers...)
+
+	// var buf bytes.Buffer
+
+	// log.Infof("Copying output from export archive readers")
+	// if w, err := io.Copy(&buf, finalTarReader); err != nil {
+	// 	log.Infof("Error from reading export archive - %s", err.Error())
+	// } else {
+	// 	log.Infof("%d bytes were read from export archive - %s", w, buf.String())
+	// }
+
+	return ioutil.NopCloser(finalTarReader), nil
+}
+
+func (c *Container) onlineContainerArchivePath(vc *viccontainer.VicContainer, path string) (io.ReadCloser, error) {
+	return nil, nil
+}
+
+// ContainerCopy performs a deprecated operation of archiving the resource at
+// the specified path in the container identified by the given name.
+func (c *Container) ContainerCopy(name string, res string) (io.ReadCloser, error) {
+	return nil, fmt.Errorf("%s does not yet implement ContainerCopy", ProductName())
+}
+
+// ContainerExport writes the contents of the container to the given
+// writer. An error is returned if the container cannot be found.
+func (c *Container) ContainerExport(name string, out io.Writer) error {
+	return fmt.Errorf("%s does not yet implement ContainerExport", ProductName())
+}
+
+// ContainerExtractToDir extracts the given archive to the specified location
+// in the filesystem of the container identified by the given name. The given
+// path must be of a directory in the container. If it is not, the error will
+// be ErrExtractPointNotDirectory. If noOverwriteDirNonDir is true then it will
+// be an error if unpacking the given content would cause an existing directory
+// to be replaced with a non-directory and vice versa.
+func (c *Container) ContainerExtractToDir(name, path string, noOverwriteDirNonDir bool, content io.Reader) error {
+	defer trace.End(trace.Begin(name))
+
+	vc := cache.ContainerCache().GetContainer(name)
+	if vc == nil {
+		return NotFoundError(name)
+	}
+
+	rawReader, err := archive.DecompressStream(content)
+	if err != nil {
+		log.Errorf("Input tar stream to ContainerExtractToDir not recognized: %s", err.Error())
+		return StreamFormatNotRecognized()
+	}
+	tarReader := tar.NewReader(rawReader)
+
+	err = c.offlineContainerExtractToDir(vc, path, tarReader)
+	if err != nil && IsResourceInUse(err) {
+		err = c.onlineContainerExtractToDir(vc, path, tarReader)
+	}
+
+	return err
+}
+
+func (c *Container) offlineContainerExtractToDir(vc *viccontainer.VicContainer, path string, tarReader *tar.Reader) error {
+	mounts := mountsFromContainer(vc)
+	mounts = append(mounts, types.MountPoint{Destination: "/"})
+	writerMap := NewArchiveStreamWriterMap(mounts, path)
+
+	for {
+		header, err := tarReader.Next()
+		if err != nil || err == io.EOF {
```

Comment:
Minor: are we not caring about non-EOF errors here? Looks like we break in either case and don't report.